### PR TITLE
tests: Add support for lazyfs

### PR DESCRIPTION
**Pull request by serathius**
Signed-off-by: Marek Siarkowicz <siarkowicz@google.com>

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by ptabor** [comment]
@serathius: Is it still intentionally draft ?

**Comment by serathius** [comment]
Yes, I only implemented building lazyfs. I still need to configure it and implement cache clear on crash.

**Comment by serathius** [comment]
Still want to revisit this. Problem is integrating lazyfs in a user approachable way. We need to run separate lazyfs process for each etcd instance. It might require non-trivial rewrite.

**Comment by serathius** [comment]
PR is ready to review. There is one concern on my side. LazyFS uses a lot of CPU, in environment with limited resources like CI it might be too much. Will need to validates if it doesn't cause flakes.

cc @ahrtr @ptabor @jmhbnz @fuweid @chaochn47

**Comment by fuweid** [comment]
>  LazyFS uses a lot of CPU, in environment with limited resources like CI it might be too much.

I run this branch in my local with 3 cores. The failure is about ` traffic.go:91: Requiring minimal 200.000000 qps for test results to be reliable, got 147.115802 qps`. I did see the lazyfs uses more CPU.

**Comment by serathius** [comment]
Need to think how to integrate the lazyfs into the test scenarios. Should we lower the qps requirement, run separate scenarios with lower qps, or only enable on LowTraffic. Let me know if you have any suggestions.

**Comment by fuweid** [comment]
> run separate scenarios with lower qps, or only enable on LowTraffic

I am trying to enable grpc-only endpoint for the test. update it later.

**Comment by fuweid** [comment]
@serathius My devbox is kind of similar to github action vm. I run the branch with 3 CPU cores. I think the `HighTraffic.minQPS` should be changed to `120` if enableLazyfs is true. I was trying to use different block size but it doesn't help. The fsync operation might require more CPU based on my testing.

image

BTW, `compactAfterCommitBatch` becomes flaky after enable lazyfs. After trigger compact, the total number of keys isn't more enough to trigger batch commit. Hope that information is help.

**Comment by fuweid** [comment]
I didn't see any qps issue in LowTraffic or Kubernetes profiles. For the change, it looks good to me.

**Comment by fuweid** [comment]
```
2023-07-25T09:21:06.6384677Z     logger.go:130: 2023-07-25T09:21:04.266Z	INFO	Triggering failpoint	{"failpoint": "compactBeforeCommitBatch"}
2023-07-25T09:21:06.6385599Z     logger.go:130: 2023-07-25T09:21:04.266Z	INFO	Setting up gofailpoint	{"failpoint": "compactBeforeCommitBatch"}
2023-07-25T09:21:06.6386514Z     logger.go:130: 2023-07-25T09:21:04.272Z	INFO	Triggering gofailpoint	{"failpoint": "compactBeforeCommitBatch"}
2023-07-25T09:21:06.6387728Z /home/runner/work/etcd/etcd/bin/etcd (TestRobustnessKubernetesLowTrafficClusterOfSize1LazyFS-test-0) (11989): {"level":"info","ts":"2023-07-25T09:21:04.390478Z","caller":"mvcc/index.go:194","msg":"compact tree index","revision":73}
2023-07-25T09:21:06.6389417Z /home/runner/work/etcd/etcd/bin/etcd (TestRobustnessKubernetesLowTrafficClusterOfSize1LazyFS-test-0) (11989): {"level":"info","ts":"2023-07-25T09:21:04.390729Z","caller":"mvcc/kvstore_compaction.go:69","msg":"finished scheduled compaction","compact-revision":73,"took":"177.702µs","hash":2606726685}
2023-07-25T09:21:06.6391083Z /home/runner/work/etcd/etcd/bin/etcd (TestRobustnessKubernetesLowTrafficClusterOfSize1LazyFS-test-0) (11989): {"level":"info","ts":"2023-07-25T09:21:04.390754Z","caller":"mvcc/hash.go:143","msg":"storing new hash","hash":2606726685,"revision":73,"compact-revision":-1}
2023-07-25T09:21:06.6392842Z     logger.go:130: 2023-07-25T09:21:04.391Z	INFO	Waiting for member to exit	{"member": "TestRobustnessKubernetesLowTrafficClusterOfSize1LazyFS-test-0"}
```

That is flaky case I mentioned. The compact doesn't commit after enable failpoint.

**Comment by fuweid** [comment]
```
stderr: # go.etcd.io/etcd/tests/v3/framework/e2e
Error: stderr: framework/e2e/cluster.go:415:34: too many arguments in call to NewEtcdProcess
stderr: have (testing.TB, *EtcdServerProcessConfig)
stderr: want (*EtcdServerProcessConfig)
Error: stderr: framework/e2e/cluster.go:835:34: too many arguments in call to NewEtcdProcess
stderr: have (testing.TB, *EtcdServerProcessConfig)
stderr: want (*EtcdServerProcessConfig)
Error: stderr: framework/e2e/cluster.go:864:34: too many arguments in call to NewEtcdProcess
stderr: have (testing.TB, *EtcdServerProcessConfig)
stderr: want (*EtcdServerProcessConfig)
Error: stderr: framework/e2e/cluster_proxy.go:43:9: cannot use NewProxyEtcdProcess(cfg) (value of type *proxyEtcdProcess) as type EtcdProcess in return statement:
stderr: *proxyEtcdProcess does not implement EtcdProcess (missing LazyFS method)
Error: stderr: framework/e2e/cluster_proxy.go:47:34: not enough arguments in call to NewEtcdServerProcess
stderr: have (*EtcdServerProcessConfig)
stderr: want (testing.TB, *EtcdServerProcessConfig)
```
need to support proxy part.

**Comment by serathius** [comment]
> need to support proxy part.

I know :(. I always neglect that.

File: Makefile [hunk 1]
```diff
@@ -149,6 +149,20 @@ ifeq (, $(shell which yamlfmt))
...
 
+# Tools
+
+.PHONY: install-lazyfs
+install-lazyfs: bin/lazyfs
+
+bin/lazyfs:
+	rm /tmp/lazyfs -rf
+	git clone https://github.com/dsrhaslab/lazyfs /tmp/lazyfs
+	cd /tmp/lazyfs/; git checkout 94ef5e60117f2a6c6d12b29e09e287c3893150ca
```
Comment (jmhbnz) [inline Makefile:160]: Could we refer to a specific tag here to be a bit more human readable?

Alternatively we could potentially just clone the specific tag so no separate checkout command is required.

**Review by jmhbnz — COMMENTED** [review]
Nice work getting this going team!

Some feedback below, nothing too major.

File: Makefile (hunk 1 above)
Comment (serathius) [reply Makefile:160]: There were not recent tags when I wrote the PR.

However, last week there was a new release, so there is a tag for us to use!

**Review by fuweid — APPROVED** [review]
LGTM!

File: Makefile [hunk 2]
```diff
@@ -149,6 +149,20 @@ ifeq (, $(shell which yamlfmt))
...
+	cd /tmp/lazyfs/libs/libpcache; ./build.sh
+	cd /tmp/lazyfs/lazyfs; ./build.sh
+	mkdir -p ./bin
+	cp /tmp/lazyfs/lazyfs/build/lazyfs ./bin/lazyfs
+
 # Cleanup
 
 clean:
```
[the remark this answers is not in the document]
Comment (serathius) [reply Makefile:168]: Done.

File: tests/framework/e2e/lazyfs.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
+// Copyright 2017 The etcd Authors
```
[the remark this answers is not in the document]
Comment (serathius) [reply tests/framework/e2e/lazyfs.go:1]: Done

File: tests/framework/e2e/lazyfs.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+func (fs *LazyFS) Start(ctx context.Context) (err error) {
+	if fs.ep != nil {
+		return nil
+	}
+	err = os.WriteFile(fs.configPath(), fs.config(), 0666)
+	if err != nil {
+		return err
+	}
+	dataPath := filepath.Join(fs.LazyFSDir, "data")
+	err = os.Mkdir(dataPath, 0777)
```
[the remark this answers is not in the document]
Comment (serathius) [reply tests/framework/e2e/lazyfs.go:58]: to cd into it.

File: tests/framework/e2e/etcd_process.go [hunk 5]
```diff
@@ -96,22 +98,29 @@ type EtcdServerProcessConfig struct {
...
-func NewEtcdServerProcess(cfg *EtcdServerProcessConfig) (*EtcdServerProcess, error) {
+func NewEtcdServerProcess(t testing.TB, cfg *EtcdServerProcessConfig) (*EtcdServerProcess, error) {
 	if !fileutil.Exist(cfg.ExecPath) {
 		return nil, fmt.Errorf("could not find etcd binary: %s", cfg.ExecPath)
 	}
 	if !cfg.KeepDataDir {
 		if err := os.RemoveAll(cfg.DataDirPath); err != nil {
 			return nil, err
 		}
+		if err := os.Mkdir(cfg.DataDirPath, 0777); err != nil {
```
[the remark this answers is not in the document]
Comment (jmhbnz) [reply tests/framework/e2e/etcd_process.go:113]: Omg. Yeah I am being a silly goose, ignore me lol... 🤦🏻‍♂️

**Review by jmhbnz — APPROVED** [review]
LGTM

**Pull request merged**